flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -98,12 +100,44 @@ protected void cleanup() {
 	}
 
 	@Override
-	protected void run() throws Exception {
-		headOperator.run(getCheckpointLock(), getStreamStatusMaintainer());
+	protected void performDefaultAction(ActionContext context) throws Exception {
+		// Against the usual contract of this method, this implementation is not step-wise but blocking instead for
```

Comment:
Wouldn't it be better to not hijack the execution from mailbox here? It's kind of fishy this way, when one task behaves in a completely different way. Maybe it would be better to run the mailbox loop in `StreamTask` always and this `performDefaultAction` would just spawn a source thread, without joining on it?

Reply:
I was thinking about that, but then 1) how the you add the synchronization part around `run()` and 2) it would still loop around the defaultAction, but starting the thread should then be ignored? I think this is actually cleaner and more performant.

Reply:
looping around (2nd and later executions) the default action should be a no-op. I guess it could even be an empty method and starting source thread could be done in another method/place. 

> 1) how the you add the synchronization part around run()

What do you mean?

Reply:
That suggestion would need an `if` in the default action that triggers for the first invocation. Starting in another method, the question is which one and we also probably require to join back in another place then.

What I mean is, the alternative loop does two things different:
- no invocation of default action.
- running all letters inside a `synchronized` block.

So the question that remains is, how you plan to add the synchronized block to the existing mailbox loop? Personally I feel like this approach might be the cleanest and most performant, but I would be open to suggestions for a complete alternative if you have something in mind.

Reply:
Ok, If I understand this correctly we have two choices:
1. (proposed in this PR) violate the contract of this method, hijack the mailbox execution and add checkpoint lock handling here
2. keep the contract of this method, but spill the hopefully soon to be deprecated checkpointing lock handling into the `StreamTask`

?

In that case you might be right that 1. sounds better.

One last question, have you considered how does this play with an idea of dropping `SourceTasks` and just using `OneInputStreamTask` to handle them? Will it be only possible for a connectors using new source interface and we would still keep this `SourceStreamTask` as a compatibility layer? Or could we somehow move this code to `OneInputStreamTask` as well?

Reply:
I was thinking through the alternatives and they have all more downsides in my opinion, performance wise and from the required modifications.

Now the "nice" thing about this compatibility layer is that we can play kind of the same trick when we remove the `SourceStreamTask` - we should always be able to just send runnables that fireup the source thread and hijack the mailbox. Question is then where and when to emit those in a nice way, but I think technially it is possible.